src/phoebus_driver.cpp

```diff
@@ -483,51 +413,323 @@ TaskCollection PhoebusDriver::RungeKuttaStage(const int stage) {
+                             mdudt.get(), beta * dt, mc1.get());
+  }
+
+  // Fix up flux failures
+  TaskRegion &async_region_2 = tc.AddRegion(num_independent_task_lists);
+  for (int ib = 0; ib < num_independent_task_lists; ib++) {
+    auto pmb = blocks[ib].get();
+    auto &tl = async_region_2[ib];
+
+    // first make other useful containers
+    auto &base = pmb->meshblock_data.Get();
+
+    // pull out the container we'll use to get fluxes and/or compute RHSs
+    auto &sc0 = pmb->meshblock_data.Get(stage_name[stage - 1]);
+    auto &sc1 = pmb->meshblock_data.Get(stage_name[stage]);
+    using MDT = std::remove_pointer<decltype(sc0.get())>::type;
+
+    // fill in derived fields
+    auto fill_derived =
+        tl.AddTask(none, parthenon::Update::FillDerived<MeshBlockData<Real>>, sc1.get());
+
+    auto fixup =
+        tl.AddTask(fill_derived, fixup::ConservedToPrimitiveFixup<MeshBlockData<Real>>,
+                   sc1.get(), sc0.get());
+
+    auto radfixup =
+        tl.AddTask(fixup, fixup::RadConservedToPrimitiveFixup<MeshBlockData<Real>>,
+                   sc1.get(), sc0.get());
+
+    auto floors =
+        tl.AddTask(radfixup, fixup::ApplyFloors<MeshBlockData<Real>>, sc1.get());
+  }
+
+  // TODO(BRR) skip this communication and source update if no radiation
+
+  // Communicate
+  TaskRegion &sync_region_3 = tc.AddRegion(num_partitions);
+  for (int ip = 0; ip < num_partitions; ip++) {
+    auto &tl = sync_region_3[ip];
+    auto &mc1 = pmesh->mesh_data.GetOrAdd(stage_name[stage], ip);
 
     auto send_local =
-        tl.AddTask(update, parthenon::cell_centered_bvars::SendBoundBufs<local>, sc1);
+        tl.AddTask(none, parthenon::cell_centered_bvars::SendBoundBufs<local>, mc1);
+    auto send_nonlocal =
+        tl.AddTask(none, parthenon::cell_centered_bvars::SendBoundBufs<nonlocal>, mc1);
+
     auto recv_local =
-        tl.AddTask(update, parthenon::cell_centered_bvars::ReceiveBoundBufs<local>, sc1);
+        tl.AddTask(none, parthenon::cell_centered_bvars::ReceiveBoundBufs<local>, mc1);
+    auto recv_nonlocal =
+        tl.AddTask(none, parthenon::cell_centered_bvars::ReceiveBoundBufs<nonlocal>, mc1);
+
     auto set_local =
-        tl.AddTask(recv_local, parthenon::cell_centered_bvars::SetBounds<local>, sc1);
+        tl.AddTask(recv_local, parthenon::cell_centered_bvars::SetBounds<local>, mc1);
+    auto set_nonlocal = tl.AddTask(
+        recv_nonlocal, parthenon::cell_centered_bvars::SetBounds<nonlocal>, mc1);
 
-    auto recv = tl.AddTask(
-        update, parthenon::cell_centered_bvars::ReceiveBoundBufs<nonlocal>, sc1);
-    auto set = tl.AddTask(recv, parthenon::cell_centered_bvars::SetBounds<nonlocal>, sc1);
+    if (pmesh->multilevel) {
+      tl.AddTask(set_nonlocal | set_local,
+                 parthenon::cell_centered_refinement::RestrictPhysicalBounds, mc1.get());
+    }
+  }
+
+  // TODO(BRR) loop this in thoughtfully!
```

Comment:
reminder

Reply:
Now thoughtfully looped in -- I removed a whole c2p + boundary conditions setting, and TWO rounds of send/recv. Code was very redundant previously.